pkg/geo/geomfn/add_measure.go

```diff
@@ -0,0 +1,128 @@
+	"github.com/twpayne/go-geom"
+)
+
+// AddMeasure takes a LineString or MultiLineString and linearly interpolates measure values for each line.
+func AddMeasure(geometry geo.Geometry, start float64, end float64) (geo.Geometry, error) {
+	t, err := geometry.AsGeomT()
+	if err != nil {
+		return geometry, err
+	}
+
+	switch t := t.(type) {
+	case *geom.LineString:
+		newLineString, err := addMeasureToLineString(t, start, end)
+		if err != nil {
+			return geometry, err
+		}
+		return geo.MakeGeometryFromGeomT(newLineString)
+	case *geom.MultiLineString:
+		newMultiLineString, err := addMeasureToMultiLineString(t, start, end)
+		if err != nil {
+			return geometry, err
+		}
+		return geo.MakeGeometryFromGeomT(newMultiLineString)
+	default:
+		// Ideally we should return NULL here, but following PostGIS on this.
+		return geometry, errors.Newf("input geometry must be LINESTRING or MULTILINESTRING")
+	}
+}
+
+// addMeasureToMultiLineString takes a MultiLineString and linearly interpolates measure values for each component line.
+func addMeasureToMultiLineString(
+	multiLineString *geom.MultiLineString, start float64, end float64,
+) (*geom.MultiLineString, error) {
+	newMultiLineString := geom.NewMultiLineString(augmentLayoutWithM(multiLineString.Layout()))
+
+	// Create a copy of the MultiLineString with measures added to each component LineString.
+	for i := 0; i < multiLineString.NumLineStrings(); i++ {
+		newLineString, err := addMeasureToLineString(multiLineString.LineString(i), start, end)
+		if err != nil {
+			return multiLineString, err
+		}
+		err = newMultiLineString.Push(newLineString)
+		if err != nil {
+			return multiLineString, err
+		}
+	}
+
+	return newMultiLineString.SetSRID(multiLineString.SRID()), nil
+}
+
+// addMeasureToLineString takes a LineString and linearly interpolates measure values.
+func addMeasureToLineString(
+	lineString *geom.LineString, start float64, end float64,
+) (*geom.LineString, error) {
+	if lineString.Empty() {
+		return geom.NewLineString(augmentLayoutWithM(lineString.Layout())), nil
+	}
+
+	// Extract the line's current points.
+	lineCoords := lineString.Coords()
+
+	// Compute the length of the line as the sum of the distances between each pair of points.
+	// Also, fill in pointMeasures with the partial sums.
+	prevPoint := lineCoords[0]
+	lineLength := float64(0)
+	pointMeasures := make([]float64, lineString.NumCoords())
+	for i := 0; i < lineString.NumCoords(); i++ {
+		curPoint := lineCoords[i]
+		distBetweenPoints := coordNorm(coordSub(prevPoint, curPoint))
+		lineLength += distBetweenPoints
+		pointMeasures[i] = lineLength
+		prevPoint = curPoint
+	}
+
+	// Compute the measures for each point.
+	for i := 0; i < lineString.NumCoords(); i++ {
+		// Handle special case where line is zero length.
+		if lineLength == 0 {
+			pointMeasures[i] = start + (end-start)*float64(i)/float64(lineString.NumCoords()-1)
+		} else {
+			pointMeasures[i] = start + (end-start)*pointMeasures[i]/lineLength
```

Comment:
can you put brackets here and above for `(pointMeasures[i]/lineLength)`? don't trust computers to implement BODMAS correctly after "B" :P